src/main/java/com/sourcegraph/cody/autocomplete/CodyEditorFactoryListener.kt

```diff
@@ -0,0 +1,183 @@
+        return
+      }
+      val completions = instance
+      completions.clearAutocompleteSuggestions(editor)
+      if (isImplicitAutocompleteEnabledForEditor(editor) &&
+          isEditorValidForAutocomplete(editor) &&
+          !CommandProcessor.getInstance().isUndoTransparentActionInProgress) {
+        informAgentAboutEditorChange(editor)
+        val changeOffset = event.offset + event.newLength
+        if (editor.caretModel.offset == changeOffset) {
+          completions.triggerAutocomplete(
+              editor, changeOffset, InlineCompletionTriggerKind.AUTOMATIC)
+        }
+      }
+    }
+  }
+
+  companion object {
+    /**
+     * Returns true if this editor is currently open and focused by the user. Returns true if this
+     * editor is in a separate tab or not focused/selected by the user.
+     */
+    private fun isSelectedEditor(editor: Editor?): Boolean {
+      if (editor == null) {
+        return false
+      }
+      val project = editor.project
+      if (project == null || project.isDisposed) {
+        return false
+      }
+      val editorManager = FileEditorManager.getInstance(project) ?: return false
+      if (editorManager is FileEditorManagerImpl) {
+        val current = editorManager.getSelectedTextEditor(true)
+        return current != null && current == editor
+      }
+      val current = editorManager.getSelectedEditor()
+      return current is TextEditor && editor == current.editor
+    }
+
+    private fun getSelection(editor: Editor): Range? {
+      val selectionModel = editor.selectionModel
+      val selectionStartPosition = selectionModel.selectionStartPosition
+      val selectionEndPosition = selectionModel.selectionEndPosition
+      if (selectionStartPosition != null && selectionEndPosition != null) {
+        return Range()
+            .setStart(
+                Position()
+                    .setLine(selectionStartPosition.line)
+                    .setCharacter(selectionStartPosition.column))
+            .setEnd(
+                Position()
+                    .setLine(selectionEndPosition.line)
+                    .setCharacter(selectionEndPosition.column))
+      }
+      val carets = editor.caretModel.allCarets
+      if (!carets.isEmpty()) {
+        val caret = carets[0]
+        val position =
+            Position()
+                .setLine(caret.logicalPosition.line)
+                .setCharacter(caret.logicalPosition.column)
+        // A single-offset caret is a selection where end == start.
+        return Range().setStart(position).setEnd(position)
+      }
+      return null
+    }
+
+    // Sends a textDocument/didChange notification to the agent server.
+    fun informAgentAboutEditorChange(editor: Editor?, skipCodebaseOnFileOpened: Boolean = true) {
+      if (editor == null) {
+        return
+      }
+      if (editor.project == null) {
+        return
+      }
+      if (!isConnected(editor.project!!)) {
+        return
+      }
```

Comment:
Minor nitpick: isn't it better to just combine those if statements into one condition?